src/MagnumPlugins/ObjImporter/ObjImporter.cpp

```diff
@@ -115,387 +372,575 @@ ObjImporter::~ObjImporter() = default;
+            // TODO: C++ 17
+            _state->objects.emplace_back(new ObjObject);
+            object = _state->objects.back().get();
+        }
+    };
+
+    /* Create 'group' if not created by 'g' keyword */
+    auto ensureGroup = [&]{
+        if(group == nullptr) {
+            ensureObject();
+
+            // TODO: C++ 17
+            object->groups.emplace_back(new ObjGroup{*object});
+            group = object->groups.back().get();
+        }
+    };
+
+    /* Create 'meshData' if not created by 'usemtl' keyword */
+    auto ensureMeshData = [&](){
+        if(meshData == nullptr) {
+            ensureGroup();
+            meshData = &group->meshDataForMaterial(-1);
+        }
+    };
 
-    /* The first mesh doesn't have name by default but we might find it later,
-       so we need to track whether there are any data before first name */
-    bool thisIsFirstMeshAndItHasNoData = true;
-    _file->meshNames.emplace_back();
+    /* Close a section and add it to the current meshData */
+    auto finishSection = [&]{
+        if(section.data() == nullptr) {
+            /* No open section */
+            return;
+        }
 
-    while(_file->in->good()) {
-        /* The previous object might end at the beginning of this line */
-        const std::streampos end = _file->in->tellg();
+        ensureMeshData();
+
+        ObjMesh* mesh;
+        if(sectionPrimitive == 'p') {
+            if(!meshData->points) {
+                // TODO: C++ 17
+                _state->meshes.emplace_back(new ObjMesh{*meshData, MeshPrimitive::Points});
+                meshData->points = _state->meshes.back().get();
+            }
+            mesh = meshData->points;
+        } else if(sectionPrimitive == 'l') {
+            if(!meshData->lines) {
+                // TODO: C++ 17
+                _state->meshes.emplace_back(new ObjMesh{*meshData, MeshPrimitive::Lines});
+                meshData->lines = _state->meshes.back().get();
+            }
+            mesh = meshData->lines;
+        } else { /* sectionPrimitive == 'f' */
+            if(!meshData->faces) {
+                // TODO: C++ 17
+                _state->meshes.emplace_back(new ObjMesh{*meshData, MeshPrimitive::Triangles});
+                meshData->faces = _state->meshes.back().get();
+            }
+            mesh = meshData->faces;
+        }
+        if(line.data() == nullptr) {
+            /* Usually for the last line */
+            mesh->sections.push_back(section);
+        } else {
+            mesh->sections.emplace_back(section.data(), size_t(line.data()-section.data()));
+        }
+        mesh->minPrimitives += minSectionPrimitives;
+
+        section = {nullptr};
+        minSectionPrimitives = 0;
+    };
+
+    auto finishObject = [&]{
+        if(object && object->groups.empty()) {
+            /* Create dummy mesh for this object so that it gets loaded as ObjectData */
+            _state->meshlessObjects.push_back(object->name);
+        }
+    };
```

Comment:
I'm a bit unsure about these inline lambdas -- i'm personally always trying to do such cleanup code only once in the parsing loop, because otherwise the corner-case testing gets quite out of hand. But I didn't stare at the code long enough yet to see how that could be done :)